src/cloudai/schema/system/slurm/slurm_system.py

```diff
@@ -493,61 +493,99 @@ def scancel(self, job_id: int) -> None:
         """
         self.cmd_shell.execute(f"scancel {job_id}")
 
-    def update_node_states(self) -> None:  # noqa: C901
+    def update_node_states(self, squeue_output: Optional[str] = None, sinfo_output: Optional[str] = None) -> None:
         """
-        Updates the states of nodes in the Slurm system by querying the current
-        state of each node using the 'sinfo' command, and correlates this with
-        'squeue' to determine which user is running jobs on each node. The method
-        parses the output of these commands, identifies the state of nodes and the
+        Updates the states of nodes in the Slurm system by querying the current state of each node using
+        the 'sinfo' command, and correlates this with 'squeue' to determine which user is running jobs on
+        each node. This method parses the output of these commands, identifies the state of nodes and the
         users, and updates the corresponding SlurmNode instances in the system.
 
-        This method does not return any value. It updates the internal state of
-        SlurmNode instances based on the current state and user information
-        reported by 'sinfo' and 'squeue'.
+        Args:
+            squeue_output (str): The output from the squeue command, if already fetched.
+            sinfo_output (str): The output from the sinfo command, if already fetched.
         """
-        node_user_map = {}
-        squeue_command = "squeue -o '%N|%u' --noheader"
-        self.logger.debug(f"Updating node user information with command: {squeue_command}")
-        squeue_stdout, squeue_stderr = self.cmd_shell.execute(squeue_command).communicate()
-        if squeue_stderr:
-            self.logger.error(f"Error querying node user information: {squeue_stderr}")
-            return
-
-        for line in squeue_stdout.split("\n"):
-            if line.strip():
-                node_list, user = line.split("|")
-                for node in self.parse_node_list([node_list]):
-                    node_user_map[node] = user
+        if squeue_output is None or sinfo_output is None:
+            squeue_output, _ = self.fetch_command_output("squeue -o '%N|%u' --noheader")
```

Comment:
Do you think it make sense to introduce specific function like:
```py
self.get_squeue_output()
self.get_snifo_output()
```

Reply:
Yes, it makes sense. I will do.

Reply:
Updated.